datafusion/core/src/optimizer/subquery_filter_to_join.rs

```diff
@@ -0,0 +1,327 @@
+                let optimized_input = self.optimize(input, execution_props)?;
+
+                // Splitting filter expression into components by AND
+                let mut filters = vec![];
+                utils::split_conjunction(predicate, &mut filters);
+
+                // Searching for subquery-based filters
+                let (subquery_filters, regular_filters): (Vec<&Expr>, Vec<&Expr>) =
+                    filters
+                        .into_iter()
+                        .partition(|&e| matches!(e, Expr::InSubquery { .. }));
+
+                // Check all subquery filters could be rewritten
+                //
+                // In case of expressions which could not be rewritten
+                // return original filter with optimized input
+                let mut subqueries_in_regular = vec![];
+                regular_filters.iter().try_for_each(|&e| {
+                    extract_subquery_filters(e, &mut subqueries_in_regular)
+                })?;
+
+                if !subqueries_in_regular.is_empty() {
+                    return Ok(LogicalPlan::Filter(Filter {
+                        predicate: predicate.clone(),
+                        input: Arc::new(optimized_input),
+                    }));
+                };
+
+                // Add subquery joins to new_input
+                // optimized_input value should retain for possible optimization rollback
+                let mut new_input = optimized_input.clone();
+                let opt_result = subquery_filters.iter().try_for_each(|&e| match e {
+                    Expr::InSubquery {
+                        expr,
+                        subquery,
+                        negated,
+                    } => {
+                        let right_input =
+                            self.optimize(&*subquery.subquery, execution_props)?;
+                        let right_schema = right_input.schema();
+                        if right_schema.fields().len() != 1 {
+                            return Err(DataFusionError::Plan(
+                                "Only single column allowed in InSubquery".to_string(),
+                            ));
+                        };
+
+                        let right_key = right_schema.field(0).qualified_column();
+                        let left_key = match *expr.clone() {
+                            Expr::Column(col) => col,
+                            _ => return Err(DataFusionError::NotImplemented(
+                                "Filtering by expression not implemented for InSubquery"
+                                    .to_string(),
+                            )),
+                        };
+
+                        let join_type = match negated {
+                            true => JoinType::Anti,
+                            false => JoinType::Semi,
+                        };
+
+                        let schema = build_join_schema(
+                            new_input.schema(),
```

Comment:
Better to use `optimized_input.schema()` and avoid creating a `mut`

Reply:
Switched to try_fold, no mutable variables required now